Components: header cake cleanup

adambbecker opened the pull request:
Cleans up the "header cake" (`client/components/header-cake`) component & fixes #143.

# Background

Over in #143, a situation arose with the current header cake component where the "back" text would wrap oddly on small screens if displayed in certain languages. After some investigation, it turns out that this situation could occur in a variety of situations as long as the text was long enough.

There was a variety of potential tweaks that were investigated (#143) and could be made in order to handle the situation. ~~This PR proposes one of them, which aligns the title to the right of the card and allows the "back" text to always remain inline with it's accompanying chevron icon.~~

**Updated: 12/14/15**

# Changes summary

This PR proposes a solution arrived at via the discussion on #143 and comments on this PR from an earlier version of it. The overall goal (visually) is to center the title text while keeping as much of the "back" text visible as possible above `480px`. Below `480px`, the "back" text will be hidden (*note: chevron still visible*) if it is 8 characters or above.

## Component structure

The component's structure is changed slightly with the introduction of a `HeaderCakeBack` sub-component whose job is just to display the "back" text/link. That sub-component also is the one that keeps track of window width and text length in order to decide if it should render just the chevron or the chevron + text.

## Visual changes

Visually, the component shouldn't be all that different from what it currently looks like (intentionally), but should be better at displaying as much of both the title and the back text as possible. This is because of a shift in the CSS where the "corners" of the header cake are no longer tied to a distributed width, but will only take up as much space as they need (with a maximum of 1/3 the container width) to display the back text which allows the title to fill all the remaining space.

The main notable visual change could occur if the back sub-component's text is over 8 characters and the window's width is below `480px`. This currently occurs in some language's translations of the word "back" as well as two other places in english: the `devdocs/design` sub-pages & in the editor's media modals' detail screens.

### Examples

##### Domain settings

`/devdocs/design/gridicons`

![screen shot 2015-12-14 at 1 23 46 pm](https://cloud.githubusercontent.com/assets/1427136/11794512/cc007a3c-a27f-11e5-879d-169c2df2c2fc.png)

![screen shot 2015-12-14 at 1 24 56 pm](https://cloud.githubusercontent.com/assets/1427136/11794588/2610a8e4-a280-11e5-8624-ba2a811d63d2.png)

##### Devdocs sub-page

`/devdocs/design/gridicons`

![screen shot 2015-12-14 at 1 24 10 pm](https://cloud.githubusercontent.com/assets/1427136/11794608/3eb6eade-a280-11e5-9da6-a5e406cee44e.png)

![screen shot 2015-12-14 at 1 24 20 pm](https://cloud.githubusercontent.com/assets/1427136/11794611/42430926-a280-11e5-88eb-c10a091f0997.png)

*Note: "back" text does away below `480px` because the string `all components` is over 8 characters*

##### Delete site

`/settings/delete-site/*`

![screen shot 2015-12-14 at 4 34 31 pm](https://cloud.githubusercontent.com/assets/1427136/11794690/9c5c9c7e-a280-11e5-8cd6-0f93fa958f8a.png)

![screen shot 2015-12-14 at 4 34 42 pm](https://cloud.githubusercontent.com/assets/1427136/11794696/aac9e6cc-a280-11e5-94b5-e662d77409d8.png)

##### Editor media modal

`/post/*` (media modal detail)

![screen shot 2015-12-14 at 4 28 00 pm](https://cloud.githubusercontent.com/assets/1427136/11794659/7c7b7f06-a280-11e5-88c6-eec2ddfee8fd.png)

![screen shot 2015-12-14 at 4 28 13 pm](https://cloud.githubusercontent.com/assets/1427136/11794661/7fe2e472-a280-11e5-975b-68fdaeede4dc.png)

## Effected implementors

The editor media modal header cake (`client/post-editor/media-modal/detail`) was the only component that initially required some tweaking, but upon further investigation, a slight refactor of the css allowed for a little cleaner implementation of the header cake itself.

Currently, the media modal header cake does some [overriding css](https://github.com/Automattic/wp-calypso/blob/a7478174b0441760c332fcb4b8c454f5b8d73dfc/client/post-editor/media-modal/detail/_style.scss#L1-L19) in order to achieve it's intended display method. With the changes to the header cake component in this PR, the media modal implementation was tweaked via cae152c so that the input no longer relies on absolute positioning in order to achieve the desired effect.

# Testing

Visit any page that contains the `HeaderCake` component and check all the text (title & back link) will no longer wrap at certain sizes. It will try to display as much text as possible, but might become ellipsis'd in certain situations.

If the "back" text is 8 characters or above and the screen size is below `480px`, only the chevron should be visible.

---

/cc @mtias, @hoverduck

sirbrillig commented:
FWIW it looks a little awkward in the Verticals Survey component:

<img width="631" alt="screen shot 2015-12-11 at 5 12 23 pm" src="https://cloud.githubusercontent.com/assets/2036909/11756784/676bbd04-a02a-11e5-8083-a0f21a872015.png">

Maybe there could be an option to center the text?

folletto commented:
Thanks for tackling this. :)

I find the right alignment however a bit odd, doesn't feel like a title in general, and more specifically consolidated mobile patterns have action items on that kind of UI element, so I feel that they could be perceived as action (iOS specifically, but Android has icons there too).

This is at the same time a very very important navigation component and a hard one. It's very important because it's what allows our sections to drill-down without complicating further the top level structure, but the bar itself is strange because in practice the title is there just to balance the white on the side of "Back", otherwise could really be removed.

My take here would be:

1. For the scope of this PR, let's keep the title centered.
2. Let's see if we can find some way to solve that "Back" issue more elegantly, if possible.

adambbecker commented:
@sirbrillig & @folletto: Thanks for the feedback! I had a feeling the right aligned text might just "not look right" and apparently it does.

For other solutions as you've both mentioned, I've outlined a few options we can take over in #143. Basically what it comes down to is that the right aligned option provides the most amount of visible text that doesn't wrap. There are also a variety of other solutions mentioned in that issue (#143), and as with most things on the web, they all come with trade-offs.

In the meantime, I'll re-center the title and ellipsis everything else. The results of which have examples in #143.

folletto commented:
> (#143) ... 2. Ellipsis "back" text

I think we should consider something along these lines, even dropping text entirely on smaller sizes (if there was a nice way to do that with different text lengths would be amazing). 

In the end, just the back arrow is a consolidated behaviour, even more on mobile, so I don't think that losing the text would be a problem. :)

adambbecker commented:
> In the end, just the back arrow is a consolidated behaviour, even more on mobile, so I don't think that losing the text would be a problem. :)

Ooh, this is good idea and agree. I'll check that out.

adambbecker commented:
@sirbrillig, @mtias, @folletto: Alrighty then, thanks for the all feedback and suggestions. I've reworked the component in order to achieve the centered title, but also displaying as much of the "back" text as possible. It also takes up @folletto's [suggestion](https://github.com/Automattic/wp-calypso/pull/1514#issuecomment-164459830) and will not show the "back" text if it is above 8 characters and the screen size is below `480px`. I've gone ahead and updated this PR's [description](https://github.com/Automattic/wp-calypso/pull/1514#issue-121776380) with an explanation of all the updates.

Also note for @mtias & @folletto that I did include a slight rework of the `client/post-editor/media-modal/detail` component's css which removes the display overrides and absolute positioning tweaks. Visually it should behave very similarly as what's currently in production, just without having to rely on the positioning tweaks.

folletto commented:
Design wise, :+1:

sirbrillig commented:
Tested on Verticals. :+1:

gwwar commented on client/components/header-cake/back.jsx:
```diff
@@ -0,0 +1,48 @@
+	mixins: [ ObserveWindowSizeMixin ],
+
+	propTypes: {
+		onClick: PropTypes.func,
```
Is this optional? If not we can add a `PropTypes.func.isRequired` here.

adambbecker replied:
In this particular case it is actually optional because of the spacer usage. In order to give the title the most about of space possible and still keep in centered, the corners were changed to display the same text in both places (the right one is hidden). The right corner still uses the back component, but doesn't include any `onClick` props.

dmsnell replied:
`onClick` was updated in the documentation above to be "required" but by not having it required here it becomes confusing. I would recommend we make the two sources of information match

gwwar commented on client/components/header-cake/back.jsx:
```diff
@@ -0,0 +1,48 @@
+
+	render() {
+		const { text, href, onClick } = this.props;
+		const hideText = window.innerWidth <= 480 && text.length >= 8;
```
I would suggest moving `480` and `8` values into named constants for readability.

gwwar commented:
Code looks good :+1: 

I found a minor issue with the loading placeholders:
<img width="618" alt="loading" src="https://cloud.githubusercontent.com/assets/1270189/11855158/6691b9de-a3fd-11e5-8c1a-16af0ec6bc6e.png">

adambbecker commented:
@gwwar: Thanks for all the feedback and taking a look!

I know you're currently occupied elsewhere (have fun btw!), just wanted to update this thread that I've updated that domain management placeholder to be more visually appropriate.

![screen shot 2015-12-17 at 1 06 29 pm](https://cloud.githubusercontent.com/assets/1427136/11877715/056ca75a-a4bf-11e5-8cd5-584bde36014d.png)

/cc @breezyskies 

---

Overall, I think we're looking pretty good here. I'll try to get one more quick review just to confirm and we'll get this out.

dmsnell commented on client/components/header-cake/back.jsx:
```diff
@@ -0,0 +1,56 @@
+
+	getDefaultProps() {
+		return {
+			text: 'Back'
```
Don't forget to translate :smile:

```js
getDefaultProps() {
	return { text: this.translate( 'Back' ) };
}
```

dmsnell commented on client/components/header-cake/back.jsx:
```diff
@@ -0,0 +1,56 @@
+				<Gridicon icon="chevron-left" size={ 18 } />
+				{ hideText ?
+					null
+				: <span className="header-cake__back-text">{ text }</span> }
```
This looks a bit off here with the `null` and the alignment. We have been using `&&` when we only have one display condition. It might help here.

```js
{ ! hideText && <span className="header-cake__back-text">{ text }</span> }
```

Of course, then it might make more sense to name the condition in the positive…

```js
const showText = window.innerWidth > HIDE_BACK_CRITERIA.windowWidth && text.length < HIDE_BACK_CRITERIA.characterLength;
```

dmsnell commented on client/components/header-cake/index.jsx:
```diff
@@ -1,35 +1,34 @@
-	render: function() {
-		var classes = classNames(
+	render() {
+		const backText = this.props.backText || this.translate( 'Back' );
```
Might want to take advantage of the soft-failing destructuring assignment here, maybe.

```js
const { backText = this.translate( 'Back' ) } = this.props;
```

dmsnell commented on client/components/header-cake/index.jsx:
```diff
@@ -40,16 +39,16 @@ module.exports = React.createClass( {
 		return (
 			<Card className={ classes }>
 				<div className="header-cake__corner">
```
What's the purpose of this `<div />`? Do we think we might add to it in the future? If not, should we collapse it into `<HeaderCakeBack />` and leave it out here?

dmsnell commented on client/components/header-cake/index.jsx:
```diff
@@ -40,16 +39,16 @@ module.exports = React.createClass( {
-				<div className="header-cake__corner" />
+				</div>
+				<div className="header-cake__corner is-spacer">
+					<HeaderCakeBack text={ backText } />
```
What is this doing? It looks like it's invisible, but I never see it appear in the code.

dmsnell commented:
@adambbecker nice looking code! I left a few comments and questions - not sure I understand all the context for this though so if they seem off-base that's my fault.

artpi commented:
Is it supposed to be like this?
![](https://cldup.com/1zlfoE9klZ-1200x1200.png)
`http://calypso.localhost:3000/domains/manage/piszek.com/edit/piszek.com`

adambbecker commented:
@dmsnell: Cool, thanks for taking a look! I took care of most of your suggestions and your context was spot on, all good stuff. Especially like the defaults for restructuring, very helpful to know!

@artpi: It is supposed to look like that to a point. It is indeed supposed to ellipsis certain texts and the component will hide the "back" text if both these requirements are met:

* "back" text character length is over `8` characters
* window width is less than `480px`

That being said, your screenshot highlighted a rather interesting situation where a screen is `275px` wide. I don't know of any devices around that size, but I did alter the "back" text hiding condition nonetheless to take care of it and it will now hide the text if the window is under `300px` no matter what.

Thanks for the feedback!

apeatling commented:
@adambbecker I think we can merge these changes now?

gwwar commented:
I retested this and it looks good! :shipit:
